metrics/prometheus.go

```diff
@@ -1771,50 +1779,52 @@ func NewPrometheusCollector(i infoProvider, f ContainerLabelsFunc, includedMetri
 	return c
 }
 
-var (
-	versionInfoDesc = prometheus.NewDesc("cadvisor_version_info", "A metric with a constant '1' value labeled by kernel version, OS version, docker version, cadvisor version & cadvisor revision.", []string{"kernelVersion", "osVersion", "dockerVersion", "cadvisorVersion", "cadvisorRevision"}, nil)
-	startTimeDesc   = prometheus.NewDesc("container_start_time_seconds", "Start time of the container since unix epoch in seconds.", nil, nil)
-	cpuPeriodDesc   = prometheus.NewDesc("container_spec_cpu_period", "CPU period of the container.", nil, nil)
-	cpuQuotaDesc    = prometheus.NewDesc("container_spec_cpu_quota", "CPU quota of the container.", nil, nil)
-	cpuSharesDesc   = prometheus.NewDesc("container_spec_cpu_shares", "CPU share of the container.", nil, nil)
-)
+func (c *PrometheusCollector) SetOpts(opts v2.RequestOptions) {
+	c.opts = opts
+}
+
+//// Describe describes all the metrics ever exported by cadvisor. It
+//// implements prometheus.PrometheusCollector.
+//func (c *PrometheusCollector) Describe(ch chan<- *prometheus.Desc) {
+//	panic("unpexected")
+//	//c.errors.Describe(ch)
+//	for _, cm := range c.containerMetrics {
+//		ch <- cm.desc(nil, []string{})
+//	}
+//}
+
+// Collect fetches latest statistics about containers and exposes them to Prometheus handler that converts them
+// to expected Prometheus format.
+// Based on option it
+// * ask the latest from provided services.
+// * returns cached results if until time expires.
+//
+// It implements prometheus.rawCollector.
+func (c *PrometheusCollector) Collect() []*dto.MetricFamily {
```

Comment:
Right. This is because we need to use this collector in new `promhttp.TransactionalHandler` similar how i implemented it on cadvisor. Looks like this has to be changed on kubelet too

Reply:
I did a bunch of digging and the fix won't be quite as trivial or straightforward as it was in this PR. This is because in k8s interaction with Prometheus client libraries goes through the component-base metrics libraries, and it's not clear to me what the best way forward is. I pinged @logicalhan to get some feedback.

xref:
- https://github.com/kubernetes/kubernetes/blob/70e7876bef5815284b43bb1fe20b55453899d721/pkg/kubelet/server/server.go#L360
- https://github.com/kubernetes/kubernetes/blob/47c63a39ed6866c1dc45eb9ed6c4a78fec8474c6/staging/src/k8s.io/component-base/metrics/registry.go#L114-L116

Annoyingly, while it would be nice to stop using k8s' deprecated `RawMustRegister`, it looks like that is still used in a couple of other places, so we can't just get rid of it and define something better for this use case:

https://cs.k8s.io/?q=RawMustRegister&i=nope&files=&excludeFiles=&repos=kubernetes/kubernetes